Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp

```diff
@@ -0,0 +1,244 @@
+            min_x_seen = point.x();
+        if (point.x() > max_x_seen)
+            max_x_seen = point.x();
+        if (point.y() < min_y_seen)
+            min_y_seen = point.y();
+        if (point.y() > max_y_seen)
+            max_y_seen = point.y();
+    }
+
+    // We create a bitmap that is bigger by 1 pixel on each side (+2) and need to account for the 0 indexed
+    // pixel positions (+1) so we make the bitmap size the delta of x/y min/max + 3
+    auto polygon_bitmap_or_error = Gfx::Bitmap::try_create(Gfx::BitmapFormat::BGRA8888, { (max_x_seen - min_x_seen) + 3, (max_y_seen - min_y_seen) + 3 });
+    if (polygon_bitmap_or_error.is_error())
+        return;
+
+    auto polygon_bitmap = polygon_bitmap_or_error.release_value();
+
+    auto polygon_painter = Gfx::Painter(polygon_bitmap);
+
+    // We want to paint the polygon into the bitmap such that there is an empty 1px border all the way around it
+    // this ensures that we have a known pixel (0,0) that is outside the polygon. Since the coordinates are relative
+    // to the layer but the bitmap is cropped to the bounding rect of the polygon we need to offset our
+    // points by the the negative of min x/y. And because we want a 1 px offset to the right and down, we + 1 this.
+    auto polygon_bitmap_delta = Gfx::IntPoint(-min_x_seen + 1, -min_y_seen + 1);
+    for (size_t i = 0; i < m_polygon_points.size() - 1; i++) {
+        polygon_painter.draw_line(m_polygon_points.at(i).translated(polygon_bitmap_delta), m_polygon_points.at(i + 1).translated(polygon_bitmap_delta), Color::Black);
+    }
+    polygon_painter.draw_line(m_polygon_points.at(m_polygon_points.size() - 1).translated(polygon_bitmap_delta), m_polygon_points.at(0).translated(polygon_bitmap_delta), Color::Black);
+
+    // Delta to use for mapping the bitmap back to layer coordinates. -1 to account for the right and down offset.
+    auto bitmap_to_layer_delta = Gfx::IntPoint(min_x_seen + m_editor->active_layer()->location().x() - 1, min_y_seen + m_editor->active_layer()->location().y() - 1);
+    flood_polygon_selection(polygon_bitmap, bitmap_to_layer_delta);
+}
+void PolygonalSelectTool::on_mousedown(Layer*, MouseEvent& event)
+{
+    auto& image_event = event.image_event();
+    if (image_event.button() != GUI::MouseButton::Primary)
+        return;
+    if (!m_selecting) {
+        m_polygon_points.clear();
+        m_last_selecting_cursor_position = event.layer_event().position();
+    }
+
+    m_selecting = true;
+
+    auto new_point = event.layer_event().position();
+
+    // this point matches the first point exactly. consider this polygon finished
+    if (m_polygon_points.size() > 0 && new_point == m_polygon_points.at(0)) {
+        m_selecting = false;
+        m_editor->image().selection().end_interactive_selection();
+        process_polygon();
+        m_editor->did_complete_action(tool_name());
+        m_editor->update();
+        return;
+    }
+
+    // prevent duplicate points back to back
```

Comment:
Super pedantic nit: This comment and a few don't match the serenity style (proper sentences,  capitalization, full-stops)

Reply:
Fixed!